xc/xenops_server_xen.ml

```diff
@@ -2723,68 +2746,92 @@ module VIF = struct
+            let di = Xenctrl.domain_getinfo xc device.frontend.domid in
+            ignore (run !Xc_resources.setup_pvs_proxy_rules [action; "vif"; vif_interface_name;
+                                                             private_path; hotplug_path]);
+            if di.Xenctrl.hvm_guest then
+              try
+                ignore (run !Xc_resources.setup_pvs_proxy_rules [action; "tap"; tap_interface_name;
+                                                                 private_path; hotplug_path])
+              with _ ->
+                (* There won't be a tap device if the VM has PV drivers loaded. *)
+                ()
+          in
+          if proxy = None then begin
+            setup "remove";
+            Xs.transaction xs (fun t ->
+                let keys = t.Xs.directory private_path in
+                List.iter (fun key ->
+                    if String.startswith pvs_proxy_key_prefix key then
+                      t.Xs.rm (Printf.sprintf "%s/%s" private_path key)
+                  ) keys
+              )
+          end else begin
+            Xs.transaction xs (fun t ->
+                t.Xs.writev private_path (xenstore_of_pvs_proxy proxy)
+              );
+            setup "add"
+          end
       )
 
   let get_state vm vif =
     with_xc_and_xs
-      (fun xc xs ->
+      (fun xc xs -> 
          try
-           let (d: Device_common.device) = device_by_id xc xs vm Device_common.Vif Newest (id_of vif) in
-           let path = Device_common.kthread_pid_path_of_device ~xs d in
-           let kthread_pid = try xs.Xs.read path |> int_of_string with _ -> 0 in
-           let pra_path = Hotplug.vif_pvs_rules_active_path_of_device ~xs d in
-           let pvs_rules_active = try (ignore (xs.Xs.read pra_path); true) with _ -> false in
-           (* We say the device is present unless it has been deleted
-              					   from xenstore. The corrolary is that: only when the device
-              					   is finally deleted from xenstore, can we remove bridges or
-              					   switch configuration. *)
-           let domid = d.Device_common.frontend.Device_common.domid in
-           let device = "vif" ^ (string_of_int domid) ^ "." ^ (string_of_int vif.position) in
-           {
-             Vif.active = true;
-             plugged = true;
-             media_present = true;
-             kthread_pid = kthread_pid;
-             device = Some device;
-             pvs_rules_active = pvs_rules_active;
-           }
+            let (d: Device_common.device) = device_by_id xc xs vm Device_common.Vif Newest (id_of vif) in
+            let domid = d.Device_common.frontend.Device_common.domid in
+            let device = "vif" ^ (string_of_int domid) ^ "." ^ (string_of_int vif.position) in
+            match vif.backend with
+            | Network.Sriov pci ->
+              let open Xenops_interface.Pci in
+              let pci' = {
+                id = "", "";
```

Comment:
You're getting away with leaving these fields "empty" because it is only the `address` field that is used by `PCI.get_state`... Please instead split off a function that just takes an address as argument from `PCI.get_state`, and call that from both `get_state` functions.

Reply:
Fixed.